### PR TITLE
Change proposer cache overwrite logic

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**

Comment in `ProposerPayloadIDsCache.SetProposerAndPayloadIDs` says:
> Ok to overwrite if the slot is already set but the payload ID is not set.

The mentioned payload ID refers to the cached ID, not the new one. The current comparison verifies the new ID.

### CODE REVIEW

File: beacon-chain/cache/payload_id.go [hunk 1]
```diff
@@ -49,14 +50,16 @@ func (f *ProposerPayloadIDsCache) SetProposerAndPayloadIDs(slot types.Slot, vId
...
-	copy(bytes[:], append(vIdBytes[:], pId[:]...))
+	var bs [vpIdsLength]byte
+	copy(bs[:], append(vIdBytes[:], pId[:]...))
 
-	_, ok := f.slotToProposerAndPayloadIDs[slot]
-	// Ok to overwrite if the slot is already set but the payload ID is not set.
+	ids, ok := f.slotToProposerAndPayloadIDs[slot]
+	// Ok to overwrite if the slot is already set but the cached payload ID is not set.
+	// Ok to overwrite if the slot is already set but the cached payload ID is not set.
 	// This combats the re-org case where payload assignment could change the epoch of.
```
Comment (rkapka) [inline beacon-chain/cache/payload_id.go:59]: There is something missing at the end of the sentence @terencechain. What should it say?

File: beacon-chain/cache/payload_id.go [hunk 2]
```diff
@@ -49,14 +50,16 @@ func (f *ProposerPayloadIDsCache) SetProposerAndPayloadIDs(slot types.Slot, vId
...
-	var bytes [vpIdsLength]byte
-	copy(bytes[:], append(vIdBytes[:], pId[:]...))
+	var bs [vpIdsLength]byte
+	copy(bs[:], append(vIdBytes[:], pId[:]...))
 
-	_, ok := f.slotToProposerAndPayloadIDs[slot]
-	// Ok to overwrite if the slot is already set but the payload ID is not set.
+	ids, ok := f.slotToProposerAndPayloadIDs[slot]
+	// Ok to overwrite if the slot is already set but the cached payload ID is not set.
+	// Ok to overwrite if the slot is already set but the cached payload ID is not set.
```
Comment (prestonvanloon) [inline beacon-chain/cache/payload_id.go:58]: Duplicate lines?
```suggestion
	// Ok to overwrite if the slot is already set but the cached payload ID is not set.
```

File: beacon-chain/cache/payload_id.go (hunk 1 above)
Comment (terencechain) [inline beacon-chain/cache/payload_id.go:59]:
```suggestion
	// This combats the re-org case where payload assignment could change at the start of the epoch.
```

File: beacon-chain/cache/payload_id.go [hunk 3]
```diff
@@ -49,14 +50,16 @@ func (f *ProposerPayloadIDsCache) SetProposerAndPayloadIDs(slot types.Slot, vId
...
 	// This combats the re-org case where payload assignment could change the epoch of.
-	if !ok || (ok && pId != [pIdLength]byte{}) {
-		f.slotToProposerAndPayloadIDs[slot] = bytes
+	byte8 := [8]byte{}
+	if !ok || (ok && bytes.Equal(ids[vIdLength:], byte8[:])) {
```
[the remark this answers is not in the document]
Comment (rkapka) [reply beacon-chain/cache/payload_id.go:60]: It was already covered. I improved a comment in https://github.com/prysmaticlabs/prysm/pull/11191/commits/dc553ac23bc15f656aff39787888b499483c3db6

File: beacon-chain/cache/payload_id.go [hunk 4]
```diff
@@ -49,14 +50,15 @@ func (f *ProposerPayloadIDsCache) SetProposerAndPayloadIDs(slot types.Slot, vId
...
 
-	_, ok := f.slotToProposerAndPayloadIDs[slot]
-	// Ok to overwrite if the slot is already set but the payload ID is not set.
-	// This combats the re-org case where payload assignment could change the epoch of.
-	if !ok || (ok && pId != [pIdLength]byte{}) {
-		f.slotToProposerAndPayloadIDs[slot] = bytes
+	ids, ok := f.slotToProposerAndPayloadIDs[slot]
+	// Ok to overwrite if the slot is already set but the cached payload ID is not set.
+	// This combats the re-org case where payload assignment could change at the start of the epoch.
+	byte8 := [8]byte{}
```
Comment (prestonvanloon) [inline beacon-chain/cache/payload_id.go:59]: Why do we hardcode 8 here but use `vpIdsLength` on line 54?

Comment (rkapka) [reply beacon-chain/cache/payload_id.go]: fixed

**Review by terencechain — APPROVED** [review]

**Pull request merged**